lib/astarte_vmq_plugin.ex

```diff
@@ -272,4 +288,60 @@ defmodule Astarte.VMQ.Plugin do
       {:error, {:already_started, pid}} -> pid
     end
   end
+
+  defp maybe_authorize_registration(mountpoint, username) do
+    [realm, device_id] = String.split(username, "/")
+    {:ok, decoded_device_id} = Device.decode_device_id(device_id, allow_extended_id: true)
+
+    case {exists?(realm, decoded_device_id), is_device_being_deleted?(realm, decoded_device_id)} do
+      {true, false} ->
+        {:ok, registration_modifiers(mountpoint, username)}
+
+      {true, true} ->
+        {:error, :device_deletion_in_progress}
+
+      {false, _} ->
+        {:error, :device_does_not_exist}
+    end
```

Comment:
You're manipulating booleans and `exists?` clearly has precedence, so you can use `cond` which is clearer in this case:
```elixir
cond do
  not exists?(...) -> {:error, :device_does_not_exist}
  is_device_being_deleted?(...) -> {:error, :device_deletion_in_progress}
  true -> {:ok, _stuff}
end
```